python/test/unit/fem/test_fem_pipeline.py

```diff
@@ -0,0 +1,134 @@
+        self.gdim = gdim
+
+        x_ = sp.symbols("x y z")
+
+        self.u_lambda = self.u
+        for i in range(gdim):
+            self.u_lambda = self.u_lambda.subs(self.x[i], x_[i])
+
+        self.u_lambda = sp.lambdify((x_), self.u_lambda, 'numpy')
+
+    def eval(self, *args):
+        """Evaluate function at (x, y, z), where x,y,z each are individual arrays
+        of the x, y and z coordinates.
+        """
+        return self.u_lambda(*args)
+
+    def __call__(self, x):
+        """Evaluate function at a single point (x, y, z) in space."""
+        out_u = self.u
+        for i in range(self.gdim):
+            out_u = out_u.subs(self.x[i], x[i])
+        return out_u
+
+    def laplace(self):
+        """ Compute Laplace of analytical solution to obtain source function."""
+        laplace_u = 0
+        for i in range(self.gdim):
+            laplace_u += sp.diff(self.u, self.x[i], self.x[i])
+        return laplace_u
+
+
+def boundary(x):
+    """Boundary marker """
+    condition = np.logical_or(x[:, 0] < 1.0e-6, x[:, 0] > 1.0 - 1.0e-6)
+    for dim in range(1, x.shape[1]):
+        c_dim = np.logical_or(x[:, dim] < 1.0e-6, x[:, dim] > 1.0 - 1.0e-6)
+        condition = np.logical_or(condition, c_dim)
+    return condition
+
+
+@pytest.mark.parametrize("degree", [2, 3, 4])
+@pytest.mark.parametrize("cell_type, gdim", [(CellType.interval, 1), (CellType.triangle, 2),
+                                             (CellType.quadrilateral, 2), (CellType.tetrahedron, 3),
+                                             (CellType.hexahedron, 3)])
+def test_manufactured_poisson(degree, cell_type, gdim):
+    """ Manufactured Poisson problem, solving u = Pi_{i=0}^gdim (1 - x[i]) * x[i]^(p - 1)
+    where p is the degree of the Lagrange function space. Solved on the
+    gdim-dimensional UnitMesh, with homogeneous Dirichlet boundary
+    conditions.
+
+    Compares values at each dof with the analytical solution.
+
+    """
+    if gdim == 1:
+        mesh = UnitIntervalMesh(MPI.comm_world, 10)
+    if gdim == 2:
+        mesh = UnitSquareMesh(MPI.comm_world, 15, 15, cell_type)
+    elif gdim == 3:
+        mesh = UnitCubeMesh(MPI.comm_world, 10, 10, 10, cell_type)
+    V = FunctionSpace(mesh, ("CG", degree))
+    u, v = TrialFunction(V), TestFunction(V)
+    x = SpatialCoordinate(mesh)
+
+    man_sol = ManufacturedSolutionPoisson(degree, gdim)
+    f = -eval(str(man_sol.laplace()))
+    a = inner(grad(u), grad(v)) * dx
+    L = inner(f, v) * dx
+
+    u_bc = Function(V)
+    with u_bc.vector.localForm() as u_local:
+        u_local.set(0.0)
+    bc = DirichletBC(V, u_bc, boundary)
+
+    uh = Function(V)
+    solve(a == L, uh, bc)
+    x = V.tabulate_dof_coordinates()
+    x_ = np.zeros((x.shape[0], 3))
+    # Fill with coordinates to work with eval
+    if gdim == 1:
+        x_[:, :1] = x
+    elif gdim == 2:
+        x_[:, :2] = x
+    else:
+        x_ = x
+
+    for i in range(V.mesh.num_cells()):
```

Comment:
Checking at every point seems a bit excessive. How abut and just a few sample points?

Reply:
Which points? We need to know which cell the points are in to do the evaluation. Computing entity collisions does not work for non-simplicies as far as I've understood. This is also related to the lax tolerance below. In certain cells, we hit the issue of small numbers, i.e.
Exact
```
[1.49250680e-71 4.05622938e-37 2.10285187e-04 2.46427953e-04
 2.26226946e-04 2.38517920e-04 2.75302632e-05 1.11055963e-04
 2.55902608e-05 9.79105632e-05 3.35024367e-06 5.17085965e-05
 3.11415872e-06 4.55879871e-05 1.40078372e-40 5.35952031e-40
 1.05333181e-04 4.90440208e-05 1.28183236e-05 1.19150421e-05]
```
Approximated:
```
[ 2.58843355e-20 -7.19762383e-21  3.91022441e-04  4.72741436e-04
  3.31029333e-04  3.42966176e-04  8.44471629e-05  1.92304646e-04
  1.09539362e-04  1.39022866e-04  1.89449749e-05  1.12650811e-04
  3.17445495e-05  6.74901978e-05  3.19225306e-21  2.03138808e-20
  1.70234919e-04  8.19984581e-05  4.04526864e-05  4.29223670e-05]
```
This fails with a smaller absolute tolerance.